Complete large milestones

damianmarti opened the pull request:
- Allow builders to complete USDC grants milestones
- Complete stage when the last milestone is paid

Steps to test a USDC grant approval, complete milestones, approve, and pay them:

- You will need 3 addresses (one builder and two admins)
- Add admin addresses to 02_deploy_large_grant_contract.ts, line 40
- Add admin addresses to admin users on seed.ts, line 60
- Deploy contracts (yarn deploy)
- Load seed data (yarn seed)
- Fund LargeGrant contract with TestToken. Go to /debug page, select TestToken and mint 10000000000 (10000 tokens, token have 6 decimals like USDC) to the LargeGrant contract address.
- Apply for a USDC grant using the builder account
- Search USDC grant at /admin page with the first admin account
- You can add some private notes
- Vote for approval
- Vote for approval with the second admin account
- Final approval with one of the 2 admin accounts.
- Go to /my-grants with the dev account
- Go to the approved grant
- Complete the first milestone.
- Go to /admin page with the first admin account
- Search completed milestone and Vote for approval.
- Go to /admin page with the second admin account
- Search completed milestone and Final Approve the milestone.
- Check grant from builder address with updated information.
- Do the same for the second milestone
- Check grant from builder address with updated information and stage marked as Completed

closes #35

damianmarti commented on packages/nextjs/hooks/pg-ens/useLargeMilestoneReview.ts:
```diff
@@ -34,7 +36,7 @@ export const useLargeMilestoneReview = (milestoneId?: number) => {
       }
 
-      if ((status === "verified" || status === "completed") && !txHash) {
+      if ((status === "verified" || status === "paid") && !txHash) {
```
Only verified and paid status needs txHash.

damianmarti commented on packages/nextjs/app/api/large-milestones/[milestoneId]/review/route.ts:
```diff
@@ -62,29 +64,45 @@ export async function POST(req: NextRequest, { params }: { params: { milestoneId
     });
 
+    // check if this is the last milestone of the stage and update the stage status to "completed"
+    if (body.status === "paid") {
+      const stage = milestone.stage;
+      const stageWithMilestones = await getStageWithMilestones(stage.id);
+      const allMilestonesPaid = stageWithMilestones
+        ? stageWithMilestones.milestones.every(m => m.status === "paid")
+        : false;
+
+      if (allMilestonesPaid) {
```
The stage should be marked as completed when all the milestones have been paid.

Pabl0cks reviewed (commented):
GJ Damu! The workflow is working nicely ♥

**A few comments related to the features from this PR:**
- The milestone status "Completed (Verified)" might only generate confusion on the user, I think it can be transparent for him and still show "Completed (Pending Review)"
- On proposed grants, showing the Proposed status on milestones might be a bit redundant (not a big deal either)
- As a builder, when you Complete a milestone (click "Submit" in the modal), is showing a "Submitting review" notification, we could change it to just "Submitting" or "Submitting milestone"

**Not related to this PR:**
- Proposed deliverables data does not appear on the milestones at grant detail page. => Could create a new issue or add it to #49 
- We missing the "Planned Milestones" data of each stage into the "Submission details" from the top part of grant detail page. Can create a new issue for it
- Lots of reloads are happening if you try to go to /admin with a regular builder. Maybe is related to having disconnects from SIWE sometimes? I think it happens specially if you have more than 1 tab at the same time, even if you just use 1 of them.

I think we missing the grant/milestone rejection 2/3 threshold, and we will need to show some visual element in the card to show the current partial approvals / partial rejections (👍👎), can create an issue for it too!

rin-st reviewed (approved):
Gj Damu and thank you a lot for testing steps!

Questions:  
if milestone is verified, as a second admin I can do only final approve, but can’t do just approve. Is it expected behavior? Because for example for grant I can just approve

Do we need to add proof of completions to milestones info on grant page?

damianmarti commented:
> GJ Damu! The workflow is working nicely ♥

Thanks @Pabl0cks for the review!!

> 
> **A few comments related to the features from this PR:**
> 
> * The milestone status "Completed (Verified)" might only generate confusion on the user, I think it can be transparent for him and still show "Completed (Pending Review)"

Changed!

> * On proposed grants, showing the Proposed status on milestones might be a bit redundant (not a big deal either)

Removed!

> * As a builder, when you Complete a milestone (click "Submit" in the modal), is showing a "Submitting review" notification, we could change it to just "Submitting" or "Submitting milestone"

Changed to "Submitting milestone". This same message will be shown to admin when reviewing the milestone.

> 
> **Not related to this PR:**
> 
> * Proposed deliverables data does not appear on the milestones at grant detail page. => Could create a new issue or add it to [Small visual tweaks in Large grants detail page #49](https://github.com/BuidlGuidl/ens-pg/issues/49)

Yes, please.

> * We missing the "Planned Milestones" data of each stage into the "Submission details" from the top part of grant detail page. Can create a new issue for it

I'm not sure about it, but you can create an issue for this and I can check it.

> * I think we missing the grant/milestone rejection 2/3 threshold, and we will need to show some visual element in the card to show the current partial approvals / partial rejections (👍👎), can create an issue for it too!

Yes. The rejection is working as the ETH grants now. An admin can reject a grant or a milestone; no more rejections are needed. 

> * Lots of reloads are happening if you try to go to /admin with a regular builder. Maybe is related to having disconnects from SIWE sometimes? I think it happens specially if you have more than 1 tab at the same time, even if you just use 1 of them. Could create an issue to investigate once we tackle the whole admin/grant page issues

I'm not having this issue, but we can create an issue to investigate it.

damianmarti commented:
> Gj Damu and thank you a lot for testing steps!

Thanks @rin-st for the review!!

> 
> Questions: if milestone is verified, as a second admin I can do only final approve, but can’t do just approve. Is it expected behavior? Because for example for grant I can just approve

I implemented it in that way because I think it was a good way to do it (removing the final approval step, since when 2 admins had approved the grant, it's enough). We can think about it and make changes if we want.

> 
> Do we need to add proof of completions to milestones info on grant page?

Added!

Pabl0cks reviewed (approved):
TYSM Damu! Merging this ♥🙌

Added the pending comments as issues or as extra items to current issues to discuss in our next meeting 👌
